### PR TITLE
Fix issue where cached entry size keeps increasing when making tiny query repeatedly 

### PR DESCRIPTION
**Pull request by alvinlin123**
**What this PR does**:
See my comment at https://github.com/cortexproject/cortex/issues/3920#issuecomment-801498264 

**Which issue(s) this PR fixes**:
This issue was found as part of #3920, not sure if this PR fixes the issue though.

**Checklist**
- [X] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/querier/queryrange/results_cache_test.go [hunk 1]
```diff
@@ -386,52 +386,60 @@ func TestShouldCache(t *testing.T) {
...
+		name                    string
+		input                   Request
+		prevCachedResponse      []Extent
+		expectedUsedCachedEntry []Extent
+		expectedRequests        []Request
+		expectedCachedResponse  []Response
 	}{
-		// 1. Test a complete hit.
 		{
+			name: "Test a complete hit.",
```
Comment (pracucci) [inline pkg/querier/queryrange/results_cache_test.go:398]: Thanks for naming it! 🙏

**Review by pracucci — COMMENTED** [review]
Thanks for working on it! I reviewed the code: good job spotting the issue, but I'm not sure this is correct.

The problem is that the extends returned by `handleHit()` overwrite the cached response. Now think about running a query over "last 24h", then running the same query over "last 1h". With the changes proposed in this PR, the second query on "last 1h" would overwrite the cached extends for the "previous 23h" while we don't want to loose them.

File: .gitignore [hunk 2]
```diff
@@ -14,6 +14,10 @@ e2e_integration_test*
 active-query-tracker
 dist/
 
+# Test generated data
+pkg/alertmanager/templates
+pkg/querier/chunks_head/
```
[the remark this answers is not in the document]
Comment (alvinlin123) [reply .gitignore:19]: Agreed, I will remove this change and submit another PR to delete this directories after test.

**Comment by alvinlin123** [comment]
> Thanks for working on it! I reviewed the code: good job spotting the issue, but I'm not sure this is correct.
> 
> The problem is that the extends returned by `handleHit()` overwrite the cached response. Now think about running a query over "last 24h", then running the same query over "last 1h". With the changes proposed in this PR, the second query on "last 1h" would overwrite the cached extends for the "previous 23h" while we don't want to loose them.

Spot on, let I have another idea, will submit a an update soon.

**Comment by alvinlin123** [comment]
I decided to go with different approach to fix the issue; instead of modifying the `[]Extent` in `partition`, I made `queryrange.handleHit` to be able to handle overlapping Extent and `queryrange.matrixMerge()` be able to deal with overlapping sample points.  This approach is probably more resilient to future change in `partition`. 

One thing I am not quite certain is the following code in `queryrange.MergeResponses`

```
for _, res := range responses {
		promResponses = append(promResponses, res.(*PrometheusResponse))
		resultsCacheGenNumberHeaderValues = append(resultsCacheGenNumberHeaderValues, getHeaderValuesWithName(res, ResultsCacheGenNumberHeaderName)...)
	}
```
Is there any special handling needed if the given `responses` contains overlapping sample points?

**Review by pracucci — APPROVED** [review]
Thanks for iterating on my feedback! Your change looks good to me. I left few nits I would be glad if you could take a look. In the meanwhile I will ping another person expert on the query-frontend for a 2nd look, cause this stuff is tricky :)

**Review by owen-d — APPROVED** [review]
Nice investigation and good job on the PR. This code is very tricky (took me a while to reacquaint myself with the code). The fix looks good to me once you address @pracucci's comments. I'm looking forward to revendoring this into Loki :)

File: pkg/querier/queryrange/query_range.go [hunk 3]
```diff
@@ -372,6 +378,20 @@ func matrixMerge(resps []*PrometheusResponse) []SampleStream {
 	return result
 }
 
+func chopOffOverlapPortion(samples []cortexpb.Sample, choppingPointTs int64) []cortexpb.Sample {
```
[the remark this answers is not in the document]
Comment (alvinlin123) [reply pkg/querier/queryrange/query_range.go:381]: I like the new suggestion, will do.

File: pkg/querier/queryrange/query_range.go [hunk 4]
```diff
@@ -372,6 +378,20 @@ func matrixMerge(resps []*PrometheusResponse) []SampleStream {
...
+
+	// assuming stream.Samples is sorted by by timestamp in ascending order.
+	searchResult := sort.Search(len(samples), func(i int) bool {
```
[the remark this answers is not in the document]
Comment (alvinlin123) [reply pkg/querier/queryrange/query_range.go:384]: Agreed, better to be more defensive.

File: pkg/querier/queryrange/results_cache_test.go [hunk 5]
```diff
@@ -386,14 +390,16 @@ func TestShouldCache(t *testing.T) {
...
-	for i, tc := range []struct {
-		input                  Request
-		prevCachedResponse     []Extent
-		expectedRequests       []Request
-		expectedCachedResponse []Response
+	for _, tc := range []struct {
+		name                      string
+		input                     Request
+		prevCachedResponse        []Extent
+		expectedUpdateCachedEntry []Extent
```
[the remark this answers is not in the document]
Comment (alvinlin123) [reply pkg/querier/queryrange/results_cache_test.go:397]: This is a refactor-fail of me when I move to different approach from modifying `partition()` to modify `queryrange.matrixMerge()`. Will fix.

File: pkg/querier/queryrange/query_range.go [hunk 6]
```diff
@@ -372,6 +378,20 @@ func matrixMerge(resps []*PrometheusResponse) []SampleStream {
...
+		return samples[i].TimestampMs > choppingPointTs
+	})
+
+	if searchResult < len(samples) {
+		return samples[searchResult:]
+	}
+
+	return nil
```
[the remark this answers is not in the document]
Comment (alvinlin123) [reply pkg/querier/queryrange/query_range.go:392]: Ah yes you are right. I was under the impression that `x[len(x):]` would panic with invalid index. It turns out `x[len(x):]` would return `[]` and `x[len(x) + 1:]` would panic. Today I learned.

**Comment by alvinlin123** [comment]
Thanks for review and the comments @pracucci  and @owen-d. Yea, this is definitely tricky stuff; I took a trip down memory lane back in my college algorithm classes (10+ years ago) trying to understand these stuff, cool stuff :-) 

I address the PR comments, let me know if they look okay.

File: pkg/querier/queryrange/query_range.go [hunk 7]
```diff
@@ -381,6 +387,27 @@ func matrixMerge(resps []*PrometheusResponse) []SampleStream {
 	return result
 }
 
+// sliceSamples assumes given samples are sorted by timestamp in ascending order and
+// return a sub slice whose first element's is the smallest timestamp that is strictly
+// bigger than the given minTs. Empty slice is returned if minTs is bigger than all the
+//timestamps in samples.
```
Comment (pracucci) [inline pkg/querier/queryrange/query_range.go:393]:
```suggestion
// timestamps in samples.
```

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

**Review by pstibrany — APPROVED** [review]
🙈

**Pull request merged**